Documentation/proposals/202411-zone-aware-sharding.md

```diff
@@ -0,0 +1,227 @@
+generated if needed. This should allow enough flexibility to cover edgecases
+not anticiapted by this proposal.
+
+## Generated configuration
+
+The following configuration uses Helm templating syntax and the proposed
+configuration structure as proposed above. Functions from the [sprig](https://masterminds.github.io/sprig/)
+library are used, as they are also present in Helm.
+
+Values generated by the operator, like `shard_index` are retrieved from an
+`.operator` object.
+
+### classic mode
+
+```yaml
+  {{ $label := .spec.shardingStrategy.label | replace "/" "_" | replace "." "_" }}
+  {{ .spec.shardingStrategy.additionalRelabelConfig | toYaml }}
+  - source_labels: [ {{ default "__address__" $label }} ] 
+    separator: ;
+    modulus: {{ .spec.shards }}
+    target_label: __tmp_hash
+    replacement: $1
+    action: hashmod
+  - source_labels: [__tmp_hash]
+    separator: ;
+    regex: {{ .operator.shardIndex }}
+    replacement: $1
+    action: keep
+```
+
+### topology mode
+
+```yaml
+  {{ $labelSuffix := .spec.shardingStrategy.label | replace "/" "_" | replace "." "_" }}
+  {{ $shardsPerZone := max 1 (.spec.shards | div (len .spec.shardingStrategy.Topology)) }}
+  {{ $promZoneIdx := .operator.shardIndex | mod $shardsPerZone }}
+  
+  {{ .spec.shardingStrategy.additionalRelabelConfig | toYaml }}
+  - source_labels: [ '__meta_kubernetes_pod_label_{{ default "topology_kubernetes_io_zone" $labelSuffix }}' ]
+    separator: ;
+    regex: {{ index .spec.shardingStrategy.topology $promZoneIdx }}
+    replacement: $1
+    action: keep
+  - source_labels: [ '__address__' ] 
+    separator: ;
+    modulus: {{ $shardsPerZone }}
+    target_label: __tmp_hash
+    replacement: $1
+    action: hashmod
+  - source_labels: [__tmp_hash]
+    separator: ;
+    regex: {{ .operator.shardIndex | div $shardsPerZone }}
+    replacement: $1
+    action: keep
+```
+
+## Prometheus instance zone assignment
+
+To make sure that prometheus instances are deployed to the correct zone (of their
+assigned target), we need to generate a [node affinity](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#node-affinity)
+or a [node selector](https://kubernetes.io/docs/tasks/configure-pod-container/assign-pods-nodes/#create-a-pod-that-gets-scheduled-to-your-chosen-node).
+
+We use the same Helm templating notation as above.
+
+The following section shows how a node affinity would be implemented.
+
+```yaml
+{{ $shardsPerZone := max 1 (.spec.shards | div (len .spec.shardingStrategy.Topology)) }}
+spec:
+  affinity:
+    nodeAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+        nodeSelectorTerms:
+        - matchExpressions:
+          - key: '{{ .spec.shardingStrategy.label }}'
+            operator: 'In'
+            values:
+            - '{{ index .spec.shardingStrategy.Topology (.operator.shardIndex | div $shardsPerZone) }}'
+```
+
+The following section shows how a node selector would be implemented.
```

Comment:
node selector would have my preference because it's simpler to use.

Reply:
removed node affinity and explained why.